src/osd/OSD.cc

```diff
@@ -5842,18 +5833,9 @@ bool OSD::ms_dispatch(Message *m)
     return true;
   }
 
-  while (dispatch_running) {
-    dout(10) << "ms_dispatch waiting for other dispatch thread to complete" << dendl;
-    dispatch_cond.Wait(osd_lock);
-  }
-  dispatch_running = true;
-
   do_waiters();
   _dispatch(m);
-  do_waiters();
-
-  dispatch_running = false;
-  dispatch_cond.Signal();
+  assert(finished.empty());
```

Comment:
Same here, I don't have any reason to think that std::list::empty is safe to call with a racing write.